_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+### Spięcie całości
+Mamy już wszystkie potrzebne części logiki NgRx, z których chcemy skorzystać. Pozostaje już tylko dowiedzieć się, jak ich użyć. Pokażę sytuację, w której użytkownik klika przycisk odpowiadający za pobranie przykładowej aktywności z API, przechodząc jednocześnie przez odpowiednie miejsca w kodzie. W trakcie czytania zachęcam do spoglądania na zamieszczony wcześniej w poście diagram przepływu.
+
+1. Zacznijmy od komponentu wywołującego początkową akcję typu `Get Activity`. Budowa klasy tego komponentu może wyglądać następująco (pomijam _template_):
+```typescript
+export class ActivityApiComponent {
+
+    constructor(private store: Store<State>) {
+    }
+
+    getActivity(): void {
+        this.store.dispatch(getActivity());
+    }
+}
+```
+Jak widać w konstruktorze, wstrzykujemy obiekt Store, na którym wykonujemy metodę `dispatch` podając w jej argumencie typ akcji. Na tym kończy się odpowiedzialność komponentu.
+
+2. Wysłaną akcję przechwytuje efekt, który, zgodnie z kodem zaprezentowanym wcześniej, wysyła żądanie do API, a otrzymawszy odpowiedź przekazuje ją jako argument nowej akcji `Add Activity`.
+3. Akcja typu `Add Activity` jest przechwytywana przez reducer, który wyłuskuje z niej przekazaną z API odpowiedź i dodaje do store.
+4. W ostatniej kolejności do gry wchodzi selektor, który wykrywa zmianę store wynikającą z działania reducera. Użycie selektora w komponencie może wyglądać następująco:
+
+```typescript
+export class ActivityListComponent implements OnInit {
+
+  activities: ActivityItemModel[] = [];
+
+  constructor(private store: Store<State>) {
+  }
+
+  ngOnInit(): void {
+    this.store.select(getActivities)
+      .subscribe(activities => this.activities = activities);
+  }
+
+}
+```
+W metodzie `ngOnInit` wskazujemy, że chcemy reagować na zmiany store przy pomocy selektora `getActivities`, a następnie przy każdej takiej zmianie aktualizujemy lokalnie utrzymywaną tablicę, którą można następnie wykorzystać w ciele komponentu.
+
+Tym samym cały proces dobiega końca, a kolejne kliknięcie przycisku wywoła go od nowa.
+
+## Inne podejścia
+Warto wspomnieć, że tak, jak dla Angulara istnieje biblioteka NgRx, tak również dla pozostałych frameworków z tzw. wielkiej trójcy znajdziemy implementacje architektury Redux:
+- Redux dla Reacta
+- Vuex dla Vue
+
+Należy również zaznaczyć, że Redux nie jest jedynym sposobem na zarządzanie stanem. Można tutaj wymienić takie alternatywy, jak:
+* [MobX](https://mobx.js.org/README.html) 
+* [Cycle.js](https://cycle.js.org/)
+* [Flux](https://github.com/facebook/flux) - warto wiedzieć, że Flux był protoplastą Reduxa
+* [Apollo](https://www.apollographql.com/)
+
+## Podsumowanie
```

Comment:
czy dużo? no nie wiem. w porównaniu do trywialnej implementacji tak.
ale, w porównaniu do testowalnej, wydzielonej do dedykowanego serwisu implementacji obsługującej cache ostatnio pobranych wyników z opcją ręczne wymuszenia odświeżenia danych na żądanie... czy to nadal jest tak więcej kodu? nie sądzę. więc porównanie trivial vs trivial jasne, redux więcej. ale kto wdraża na prod taki trivial nie robiąc reszty? ;-)